hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java

```diff
@@ -90,21 +89,32 @@ public XmlConfigBuilder(URL url) throws IOException {
      * Constructs a XmlConfigBuilder that tries to find a usable XML configuration file.
      */
     public XmlConfigBuilder() {
-        XmlConfigLocator locator = new XmlConfigLocator();
-        this.in = locator.getIn();
-        this.configurationFile = locator.getConfigurationFile();
-        this.configurationUrl = locator.getConfigurationUrl();
+        this((XmlConfigLocator) null);
     }
 
     /**
-     * Gets the current used properties. Can be null if no properties are set.
+     * Constructs a {@link XmlConfigBuilder} that loads the configuration
+     * with the provided {@link XmlConfigLocator}.
+     * <p/>
+     * If the provided {@link XmlConfigLocator} is {@code null}, a new
+     * instance is created and the config is located in every possible
+     * places. For these places, please see {@link XmlConfigLocator}.
+     * <p/>
+     * If the provided {@link XmlConfigLocator} is not {@code null}, it
+     * is expected that it already located the configuration XML to load
+     * from. No further attempt to locate the configuration XML is made
+     * if the configuration XML is not located already.
      *
-     * @return the current used properties
-     * @see #setProperties(java.util.Properties)
+     * @param locator the configured locator to use
      */
-    @Override
-    public Properties getProperties() {
-        return properties;
+    public XmlConfigBuilder(XmlConfigLocator locator) {
+        if (locator == null) {
```

Comment:
There is a default constructor calling this one with `null` parameter. This constructor is introduced to support the config location logic as described in the TDD. See `HazelcastInstanceFactory#newHazelcastInstance(com.hazelcast.config.Config)`. It is not expected that the user calls this constructor. Maybe we can introduce a public helper class that actually hides this constructor, but I'm not sure we need it. The users are expected to interact with the `ClasspathXmlConfig`, `FileSystemXmlConfig` etc and with the aforementioned `HazelcastInstanceFactory#newHazelcastInstance()` with or without the already built config.

Reply:
Not a big deal (you can even leave it as it is), but normally I'd expect the constructors like this:
```java
public XmlConfigBuilder() {
        this(new XmlConfigLocator());
    }

    public XmlConfigBuilder(XmlConfigLocator locator) {
        checkNotNull(locator);
        this.in = locator.getIn();
        this.configurationFile = locator.getConfigurationFile();
        this.configurationUrl = locator.getConfigurationUrl();
    }
```

Reply:
Ouch... no excuses :)
I change this and then merge.

Reply:
Now I remember. Since we need to call `locator.locateEverywhere()` in the default constructor case, we can't do it in the suggested way. 

We can do changes like this:
- Create `ConfigLocator` interface
- Make `locateEverywhere()` to `return this` as `ConfigLocator` interface (it always returns `true` or throws exception)

I have taken it as a note and return to this together with the class hierarchy suggestion in favor of making progress now. 

Thanks for the suggestion :+1: